### PR TITLE
Limit the number of `StackItem`

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

File: neo/SmartContract/ApplicationEngine.cs [hunk 1]
```diff
@@ -280,17 +282,38 @@ private bool CheckStackSize(OpCode nextInstruction)
...
-                            size = array.Count;
-                        else
-                            return false;
+                    case OpCode.KEYS:
+                    case OpCode.VALUES:
+                        size = -1;
+                        break;
+                    case OpCode.NEWARRAY:
+                    case OpCode.NEWSTRUCT:
+                        size = (int)CurrentContext.EvaluationStack.Pop().GetBigInteger();
```
Comment (shargon) [inline neo/SmartContract/ApplicationEngine.cs]: Peek?

**Comment by shargon** [comment]
I think that maybe is faster just have a counter, and INC or DEC acording to the opcode

- if PUSH INC
- if POP, DEC

**Comment by erikzhang** [comment]
> I think that maybe is faster just have a counter, and INC or DEC acording to the opcode

This is hard. If we have an array on the stack:

```
[1, 2, 3]
```

We have a total of 4 `StackItem`s. After executing a `DUP` opcode:

```
[1, 2, 3] [1, 2, 3]
```

We have 5 `StackItem`s, because these two arrays are the same object. Then we execute a `DROP` opcode:

```
[1, 2, 3]
```

We are not sure how many `StackItem`s should be subtracted from the counter.

**Comment by erikzhang** [comment]
In the example above, how do you know if you should subtract 1 or 4?

**Comment by shargon** [comment]
In this example, if you `DUP` an array of 1024 elements, you have 20**50** items, 1024 per array, and 2 arrays, is the unique problem, with your solution, you have 1026 items.

But is more faster, think that now, for every opcode, compute the sum of ALL items

**Comment by erikzhang** [comment]
> In this example, if you `DUP` an array of 1024 elements, you have 2050 items, 1024 per array, and 2 arrays, ...

This is a good idea, but I am worried that it will be easy to reach the maximum number of items. Think about `DUP` every time you set or get an array element, which doubles the number of items.

**Comment by shargon** [comment]
We can use my solution for fast check, and if is passed my counter, use your sum?

**Comment by erikzhang** [comment]
> We can use my solution for fast check, and if is passed my counter, use your sum?

OK.

**Comment by shargon** [comment]
With my solution we need to know the parent, or we can't do that xD

`[a,[a,[a]]]`

i think that is more difficult than i thought ... maybe something like that ...

```csharp
class ExecutionEngine
{
int _counter=0;

public int counter 
{ 
  get{ return _counter; } 
  set{ _counter=value; if (value>MaxStackItems) state=FAULT; } 
}
}

public class StackItem:IDisposable
{
public StackItem(ExecutionEngine engine) { engine.Counter++ }
public void Dispose() { engine.Counter -- }
}

```

**Comment by erikzhang** [comment]
When to dispose?

**Comment by shargon** [comment]
c# is more difficult than c++ (I never thought that I will say this xDD)

Because on c++ in need to free the memory when i don't use the element (there is not Garbage collector), is for that reason that i create a `Claim`, `Unclaim` and `Free` methods, for prevent dispose items that are used in other site ... so for me is easy to know the moment of Dispose, i understand than here is harder ...

https://github.com/CityOfZion/neo-hypervm/blob/development/src/NeoVM/IClaimable.h#L5-L33

So when i free the item, i check before if was not claimed

https://github.com/CityOfZion/neo-hypervm/blob/development/src/NeoVM/IStackItem.cpp

**Comment by erikzhang** [comment]
@shargon I've optimized its performance, and now the number of stack items is recalculated only when possible to exceed the limit. Please take a look.
@PeterLinX Please test it.

**Comment by PeterLinX** [comment]
@erikzhang According to test, this version cost 50% more time to sync all over. Except this, everything is fine.

**Comment by erikzhang** [comment]
This is acceptable. I will merge this PR first.

**Comment by shargon** [comment]
@PeterLinX 50% slower this version?

**Pull request merged**